pennylane/ftqc/utils.py

```diff
@@ -0,0 +1,229 @@
+            dev = qml.device('default.qubit')
+
+            @qml.qnode(qml.device("null.qubit"), mcm_method="one-shot")
+            def circuit_mbqc(start_state, angles):
+                q_mgr = QubitMgr(num_qubits=5, start_idx=0)
+                input_idx = q_mgr.acquire_qubit()
+
+                # prep input node
+                qml.StatePrep(start_state, wires=[input_idx])
+
+                # prep and consume graph state iteratively
+                for i in range(num_iter):
+                    # Acquire 4 free qubit indices
+                    graph_wires = q_mgr.acquire_qubits(4)
+
+                    # Denote the index for the final output state
+                    output_idx = graph_wires[-1]
+
+                    # Prepare the state
+                    qml.ftqc.GraphStatePrep(lattice.graph, wires=graph_wires)
+
+                    # entangle input and graph using first qubit
+                    qml.CZ([input_idx, graph_wires[0]])
+
+                    # MBQC Z rotation: X, X, +/- angle, X
+                    # Reset operations allow qubits to be returned to the pool
+                    m0 = measure_x(input_idx, reset=True)
+                    m1 = measure_x(graph_wires[0], reset=True)
+                    m2 = cond_measure(m1, partial(measure, angle=angle, reset=True), partial(measure, angle=-angle, reset=True))(plane="XY", wires=graph_wires[1])
+                    m3 = measure_x(graph_wires[2], reset=True)
+
+                    # corrections based on measurement outcomes
+                    qml.cond((m0+m2)%2, qml.Z)(graph_wires[3])
+                    qml.cond((m1+m3)%2, qml.X)(graph_wires[3])
+
+                    # The input qubit can be freed and the output qubit becomes the next iteration's input
+                    q_mgr.release_qubit(input_idx)
+                    input_idx = output_idx
+
+                    # We can now free all but the last qubit, which has become the new input_idx
+                    q_mgr.release_qubits(graph_wires[0:-1])
+
+                # Perform the measurements on the output qubit from the last iteration
+                return qml.expval(X(output_idx)), qml.expval(Y(output_idx)), qml.expval(Z(output_idx))
+
+        For each loop iteration, the measured and reset wire labels are returned to the ``QubitMgr`` instance, which are then reallocated
+        on the next step, which when combined with the MCM resets allows for qubit index recycling.
+
+    """
+
+    def __init__(self, num_qubits: int = 0, start_idx: int = 0):
+        # All resources are protected via a re-entrant lock, to ensure exclusive access when
+        # acquiring/accessing/releasing qubits.
+        self._lock = RLock()
+        self._num_qubits = num_qubits
+        self._active = set()
+        is_valid = lambda x: (isinstance(x, int) and x >= 0)
+        if is_valid(num_qubits) and is_valid(start_idx):
```

Comment:
Wondering if `is_valid` works for both `num_qubits` and `start_idx`? There is a scenario that it would not throw errors when `num_qubits==0` and `start_idx==0`.

Reply:
I think this is fine as is. Both can be 0, or any positive integer.